flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobManagerSharedServicesTest.java

```diff
@@ -0,0 +1,161 @@
+import org.apache.flink.runtime.util.Hardware;
+import org.apache.flink.runtime.util.TestingFatalErrorHandler;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Tests auxiliary shared services created by {@link JobManagerSharedServices} and used by the
+ * {@link JobMaster}.
+ */
+public class JobManagerSharedServicesTest extends TestLogger {
+
+    private static final int NUM_FUTURE_THREADS = 8;
+    private static final int NUM_IO_THREADS = 4;
+    private static final int CPU_CORES = Hardware.getNumberCPUCores();
+
+    @Test
+    public void testFutureExecutorNoConfiguration() throws Exception {
+        Configuration config = new Configuration();
+
+        final JobManagerSharedServices jobManagerSharedServices =
+                JobManagerSharedServices.fromConfiguration(
+                        config,
+                        new BlobServer(config, new VoidBlobStore()),
+                        new TestingFatalErrorHandler());
+
+        CountDownLatch releasedLatch = new CountDownLatch(CPU_CORES);
+        CountDownLatch unreleasedLatch = new CountDownLatch(CPU_CORES + 1);
+        ScheduledExecutorService futureExecutor = jobManagerSharedServices.getFutureExecutor();
+        Runnable countsDown =
+                () -> {
+                    releasedLatch.countDown();
+                    unreleasedLatch.countDown();
+                };
+        for (int i = 0; i < CPU_CORES; i++) {
+            futureExecutor.execute(countsDown);
+        }
+        // Give the executor enough time to process all the Runnables
+        releasedLatch.await(1L, TimeUnit.SECONDS);
+        assertEquals(0, releasedLatch.getCount());
+        assertEquals(1, unreleasedLatch.getCount());
+        jobManagerSharedServices.shutdown();
+    }
+
+    @Test
+    public void testFutureExecutorConfiguration() throws Exception {
+        Configuration config = new Configuration();
+        config.setInteger(JobManagerOptions.JOB_MANAGER_FUTURE_THREADS, NUM_FUTURE_THREADS);
+
+        final JobManagerSharedServices jobManagerSharedServices =
+                JobManagerSharedServices.fromConfiguration(
+                        config,
+                        new BlobServer(config, new VoidBlobStore()),
+                        new TestingFatalErrorHandler());
+
+        CountDownLatch releasedLatch = new CountDownLatch(NUM_FUTURE_THREADS);
+        CountDownLatch unreleasedLatch = new CountDownLatch(NUM_FUTURE_THREADS + 1);
+        ScheduledExecutorService futureExecutor = jobManagerSharedServices.getFutureExecutor();
+        Runnable countsDown =
+                () -> {
+                    releasedLatch.countDown();
+                    unreleasedLatch.countDown();
+                };
```

Comment:
I think the the runnable also needs to block. Otherwise the test will also pass if the `futureExecutor` has only a single thread.

Reply:
Agreed, upon some deeper though I think we only need a single latch that blocks after counting down. We can even submit it to the same executor first and ensure we get to zero. Then create a new latch with an additional count and ensure it only goes down to 1.

Reply:
Deduplicated

Reply:
Used approach from [#17129](https://github.com/apache/flink/pull/17129) that test with both latches in the same runnable so there is less code.